aspects/generic_channel.go

```diff
@@ -0,0 +1,176 @@
+package ginmon
+
+import (
+	"bytes"
+	"encoding/gob"
+	"sort"
+	"sync"
+	"time"
+)
+
+// DataChannel is the data you pass into the channel. Using Name we
+// will put the Value into the right bucket.
+type DataChannel struct {
+	Name  string
+	Value float64
+}
+
+type dataStore struct {
+	sync.RWMutex
+	data map[string][]float64
+}
+
+func NewDataStore() dataStore {
+	return dataStore{data: make(map[string][]float64)}
+}
+
+func (ds dataStore) ResetKey(key string) {
+	ds.Lock()
+	defer ds.Unlock()
+	ds.data[key] = make([]float64, 0)
+}
+
+func (ds dataStore) Get(key string) []float64 {
+	ds.RLock()
+	defer ds.RUnlock()
+	return ds.data[key]
+}
+
+func (ds dataStore) Add(key string, value float64) {
+	ds.data[key] = append(ds.data[key], value)
+}
+
+// GenericChannelAspect, exported fields are used to store json
+// fields. All fields are measured in nanoseconds.
+type GenericChannelAspect struct {
+	gcdLock   sync.RWMutex
+	name      string
+	tempStore dataStore
+	Gcd       map[string]GenericChannelData
+}
+
+// GenericChannelData
+type GenericChannelData struct {
+	Count     int       `json:"count"`
+	Min       float64   `json:"min"`
+	Max       float64   `json:"max"`
+	Mean      float64   `json:"mean"`
+	Stdev     float64   `json:"stdev"`
+	P90       float64   `json:"p90"`
+	P95       float64   `json:"p95"`
+	P99       float64   `json:"p99"`
+	Timestamp time.Time `json:"timestamp"`
+}
+
+// NewGenericChannelAspect returns a new initialized GenericChannelAspect
+// object.
+func NewGenericChannelAspect(name string) *GenericChannelAspect {
+	gc := &GenericChannelAspect{name: name}
+	gc.tempStore = NewDataStore()
+	gc.Gcd = make(map[string]GenericChannelData, 0)
+	return gc
+}
+
+// StartTimer will call a forever loop in a goroutine to calculate
+// metrics for measurements every d ticks.
+func (gc *GenericChannelAspect) StartTimer(d time.Duration) {
+	timer := time.Tick(d)
+	go func() {
+		for {
+			<-timer
+			gc.calculate()
+		}
+	}()
+}
+
+// SetupGenericChannelAspect returns an unbuffered channel for type
+// DataChannel, such that you can send arbitrary key (string) value
+// (float64) pairs to it.
+func (gc *GenericChannelAspect) SetupGenericChannelAspect() chan DataChannel {
+	_gc := gc // save gc in closure
```

Comment:
I would rename this variable. "_" in go has a very well defined meaning and I would not mess with it.